### PR TITLE
limit with partition by

### PR DESCRIPTION
**Pull request by kvnkho**
check comments

### CODE REVIEW

File: fugue/execution/native_execution_engine.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
     EMPTY_PARTITION_SPEC,
     PartitionCursor,
     PartitionSpec,
+    _parse_presort_exp,
```
Comment (kvnkho) [inline fugue/execution/native_execution_engine.py:10]: I need this in order to parse the limit presort. I know it's bad practice to import these _functions. Should we change it?

File: fugue/execution/native_execution_engine.py [hunk 2]
```diff
@@ -279,6 +280,35 @@ def sample(
...
+        _presort: IndexedOrderedDict = presort or partition_spec.presort
+
+        if len(_presort.keys()) > 0:
+            d = d.sort_values(list(_presort.keys()), ascending=list(_presort.values()))
+
+        if len(partition_spec.partition_by) == 0:
+            d = d.head(n)
+        else:
+            d = d.groupby(by=partition_spec.partition_by, dropna=False).head(n)
+
```
Comment (kvnkho) [inline fugue/execution/native_execution_engine.py:311]: I tried to use pl_utils.safe_groupby_apply but I got errors so I did this for now and it passes the tests.

File: fugue/extensions/_builtins/processors.py [hunk 3]
```diff
@@ -233,6 +233,16 @@ def process(self, dfs: DataFrames) -> DataFrame:
...
 
+class Limit(Processor):
+    def process(self, dfs: DataFrames) -> DataFrame:
+        assert_or_throw(len(dfs) == 1, FugueWorkflowError("not single input"))
+        # All _get_or operations convert float to int
+        n = self.params.get_or_none("n", int)
+        presort = self.params.get_or_none("presort", str)
+        partition_spec = self.partition_spec
+        return self.execution_engine.limit(dfs[0], n, presort, partition_spec)
+
```
Comment (kvnkho) [inline fugue/extensions/_builtins/processors.py:244]: As discussed, everything goes down to the execution engine.

File: fugue/workflow/workflow.py [hunk 4]
```diff
@@ -609,6 +610,29 @@ def sample(
...
+            isinstance(n, int),
+            ValueError("n needs to be an integer"),
+        )
+        if presort is not None:
+            params["presort"] = presort
+        df = self.workflow.process(
+            self, using=Limit, pre_partition=self.partition_spec, params=params
+        )
+        return self._to_self_type(df)
+
```
Comment (kvnkho) [inline fugue/workflow/workflow.py:635]: Same as before, everything goes down to the execution engine.

**Comment by kvnkho** [comment]
This is complete. The Dask implementation was the hardest one, and might need a closer look. There are two cases, the first is partition and then apply Pandas `sort_values`. That is trivial. The second case is the one with no partition, but with a `presort`. In this case, my worry is that just taking the `limit` and then `presort` on the result will be inaccurate. If there are ties for the last record, the ties will be lost. Dask does not have support to keep the ties.

So here is my solution with 3 dfs.
1. The first presort is called the `primary_sort`. We first apply the primary sort and take limit (df1 - original limit) **O(n)**
_This is O(n) because we just keep track of the n smallest/largest rows_
2. Get the last row of df1. Save the value of the `primary_sort` column. Look for the rows with same value from the original df (df2 - these are the ties) **O(n)**
3. Get the rows where `primary_sort` is NULL. These are needed for NULL first or NULL last (df3 - nulls) **O(n)**

Concat these 3 dfs. Ideally the df size now is small (whatever the limit value was). We convert this to pandas and then apply the Pandas sort_value (should be a trivial **O(n)**). 

This circumvents the lack of sorting from Dask. It might also be more efficient than the global sort?

**Comment by goodwanghan** [comment]
null first null last vs null smallest and null largest, you have to make a decision
spark limit is using null smallest and does not allow you to specify
pandas use null las, and only let you specify one for all
sql, by default, use null smallest, but when you specify, you can only specify null first or last, but you can specify for each column

what a messy situation

**Comment by kvnkho** [comment]
I have thought about this. 

We should follow Pandas of allowing for NULLs first/last instead of Spark and SQL convention of NULLs as smallest value. 

I say this because Pandas NULLs first/last can achieve the ascending/descending of Spark and SQL by changing NULLs first/last. The other way around is not possible though. If you always treat NULLs as smallest value, a user cannot sort ascending and put the NULL values at the bottom.

Pandas implementation and Spark implementation of this current PR is good. The issue is the Dask implementation of this. As discussed, the proposed solution will not work well if the partitions are highly imbalanced.

**Comment by kvnkho** [comment]
I believe this is complete. The main change is the Dask implementation of `limit`. The most difficult case is when there is no partition, but a presort is defined. In this case, we need to apply the limit operation first. The first is a `map_partition` to `sort` and `take` the head on each of the default partition. After calling `compute()`, this becomes a much smaller `pandas DataFrame`. From here we can use the native `sort_values` to return what we want.

**Comment by goodwanghan** [comment]
👍

**Review by goodwanghan — APPROVED** [review]
add meta to dask, and you are good to go!

**Pull request merged**